Mathlib/CategoryTheory/Localization/CalculusOfFractions/ComposableArrowsTwo.lean

```diff
@@ -0,0 +1,51 @@
+/-
+Copyright (c) 2024 Joël Riou. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Joël Riou
+-/
+import Mathlib.CategoryTheory.ComposableArrows
+import Mathlib.CategoryTheory.Localization.CalculusOfFractions
+
+/-! # Essential surjectivity of the functor induced on tuples of composable arrows
+
+Assuming that `L : C ⥤ D` is a localization functor for a class of morphisms `W`
+that has both calculus of left and right fractions, we show in this file
+that the functor `L.mapComposableArrows 2 : ComposableArrows C 2 ⥤ ComposableArrows D 2`
+is essentially surjective.
+
+-/
+
+namespace CategoryTheory
+
+open Category
+
+namespace Localization
+
+variable {C D : Type*} [Category C] [Category D] (L : C ⥤ D) (W : MorphismProperty C)
+  [W.HasLeftCalculusOfFractions] [W.HasRightCalculusOfFractions]
+  [L.IsLocalization W]
+
+lemma essSurj_mapComposableArrows_two : (L.mapComposableArrows 2).EssSurj  where
+  mem_essImage Y := by
+    obtain ⟨Y₀, Y₁, Y₂, f, g, rfl⟩ := ComposableArrows.mk₂_surjective Y
+    have := essSurj L W
+    obtain ⟨f', hf'⟩ := exists_rightFraction L W
+      ((L.objObjPreimageIso Y₀).hom ≫ f ≫ (L.objObjPreimageIso Y₁).inv)
+    obtain ⟨g', hg'⟩ := exists_leftFraction L W
+      ((L.objObjPreimageIso Y₁).hom ≫ g ≫ (L.objObjPreimageIso Y₂).inv)
+    refine ⟨ComposableArrows.mk₂ f'.f g'.f,
+      ⟨ComposableArrows.isoMk₂
+        (Localization.isoOfHom L W f'.s f'.hs ≪≫ L.objObjPreimageIso Y₀)
+        (L.objObjPreimageIso Y₁)
+        ((Localization.isoOfHom L W g'.s g'.hs).symm ≪≫ (L.objObjPreimageIso Y₂)) ?_ ?_⟩⟩
+    · dsimp
+      rw [← cancel_mono (L.objObjPreimageIso Y₁).inv, assoc, assoc, assoc, hf',
+        Iso.hom_inv_id, comp_id, MorphismProperty.RightFraction.map_s_comp_map]
+    · dsimp [ComposableArrows.Precomp.map]
+      rw [← cancel_mono (L.objObjPreimageIso Y₂).inv, assoc, assoc, assoc, hg',
+        Iso.hom_inv_id, comp_id, IsIso.comp_inv_eq,
+        MorphismProperty.LeftFraction.map_comp_map_s]
```

Comment:
I was nerdsniped and this works for all `n`. (and `W.HasLeftCalculusOfFractions` is not needed)
```
lemma essSurj_mapComposableArrows (n) : (L.mapComposableArrows n).EssSurj  where
  mem_essImage Y := by
    have := essSurj L W
    induction n
    · obtain ⟨Y, rfl⟩ := ComposableArrows.mk₀_surjective Y
      exact ⟨.mk₀ _, ⟨ComposableArrows.isoMk₀ (L.objObjPreimageIso Y)⟩⟩
    · next n IH =>
      obtain ⟨Y, Z, f, rfl⟩ := ComposableArrows.precomp_surjective Y
      obtain ⟨Y', ⟨e⟩⟩ := IH Y
      obtain ⟨f', hf'⟩ := exists_rightFraction L W
        ((L.objObjPreimageIso Z).hom ≫ f ≫ (e.app 0).inv)
      refine ⟨Y'.precomp f'.f, ⟨NatIso.ofComponents ?_ ?_⟩⟩
      · rintro ⟨(_ | i), hi⟩
        · exact Localization.isoOfHom L W _ f'.hs ≪≫ L.objObjPreimageIso Z
        · exact e.app ⟨i, Nat.add_lt_add_iff_right.mp hi⟩
      · rintro ⟨(_ | i), hi⟩ ⟨(_ | j), hj⟩ ⟨⟨hij⟩⟩
        · simp
        · have : (L.map f'.s ≫ (L.objObjPreimageIso Z).hom) ≫ f = L.map f'.f ≫ e.hom.app 0 := by
            rwa [MorphismProperty.RightFraction.map, IsIso.eq_inv_comp, ← Category.assoc,
              ← Category.assoc, Iso.comp_inv_eq, Iso.app_hom] at hf'
          rcases j with _ | j
          · exact this.symm
          · dsimp
            conv_rhs => rw [← Category.assoc, this, Category.assoc, ← e.hom.naturality]
            rw [Functor.map_comp, Category.assoc]
            rfl
        · cases hij
        · exact e.hom.naturality _
```

Reply:
Thanks very much! I have just opened #13558 because passing from the case of right fractions to left fractions required some more work.